src/irr/asset/CGLTFLoader.cpp

```diff
@@ -249,66 +273,48 @@ namespace irr
+							bufferBinding.offset = globalOffsetInBufferBindingResource;
 
-							cpuMeshBuffer->setIndexCount(glTFAccessor.count.value());
+							idReferenceBindingBuffers[bufferBindingId] = cpuBuffers[bufferBindingId];
+							bufferBinding.buffer = idReferenceBindingBuffers[bufferBindingId];
 
 							switch (target)
 							{
 								case SGLTFBufferView::SGLTFT_ARRAY_BUFFER:
 								{
-									idReferenceVertexBuffers[bufferId] = cpuBuffers[bufferId];
-									const uint32_t bindingId = idReferenceVertexBuffers.size();
-									bufferBinding.buffer = idReferenceVertexBuffers[bufferId];
-
-									cpuMeshBuffer->setIndexType(EIT_UNKNOWN);
-									cpuMeshBuffer->setVertexBufferBinding(std::move(bufferBinding), bindingId);
+									cpuMeshBuffer->setVertexBufferBinding(std::move(bufferBinding), bufferBindingId);
 
-									vertexInputParams.enabledBindingFlags |= core::createBitmask({ bindingId });
-									vertexInputParams.bindings[bindingId].inputRate = EVIR_PER_VERTEX;
-									vertexInputParams.bindings[bindingId].stride = 16; // TODO: check it
+									vertexInputParams.enabledBindingFlags |= core::createBitmask({ bufferBindingId });
+									vertexInputParams.bindings[bufferBindingId].inputRate = EVIR_PER_INSTANCE;
+									vertexInputParams.bindings[bufferBindingId].stride = 16; // TODO: check it
 
-									vertexInputParams.enabledAttribFlags |= core::createBitmask({ queryAttributeId });
-									vertexInputParams.attributes[queryAttributeId].binding = bindingId;
-									vertexInputParams.attributes[queryAttributeId].format = format;
-									vertexInputParams.attributes[queryAttributeId].relativeOffset; // TODO
+									const auto attributeId = queryAttributeId.value();
+									vertexInputParams.enabledAttribFlags |= core::createBitmask({ attributeId });
+									vertexInputParams.attributes[attributeId].binding = bufferBindingId;
+									vertexInputParams.attributes[attributeId].format = format;
+									vertexInputParams.attributes[attributeId].relativeOffset = relativeOffsetInBufferViewAttribute;
 								} break;
 
 								case SGLTFBufferView::SGLTFT_ELEMENT_ARRAY_BUFFER:
 								{
 									// TODO: make sure glTF data has validated index type
 
-									switch (glTFAccessor.componentType.value())
-									{
-										case SGLTFAccessor::SCT_UNSIGNED_SHORT:
-										{
-											bufferBinding.buffer = core::make_smart_refctd_ptr<ICPUBuffer>(glTFAccessor.count.value() * sizeof(uint16_t));
-											cpuMeshBuffer->setIndexType(EIT_16BIT);
-										} break;
-
-										case SGLTFAccessor::SCT_UNSIGNED_INT:
-										{
-											bufferBinding.buffer = core::make_smart_refctd_ptr<ICPUBuffer>(glTFAccessor.count.value() * sizeof(uint32_t));
-											cpuMeshBuffer->setIndexType(EIT_32BIT);
-										} break;
-									}
-
+									bufferBinding.offset += relativeOffsetInBufferViewAttribute;
 									cpuMeshBuffer->setIndexBufferBinding(std::move(bufferBinding));
```

Comment:
just to be devil's advocate, does glTF always require indices? Or can have a case where there can be no index buffer (which is an implicit iota 0,1,2,3,4,... index buffer?)

Reply:
it doesn't, but the engine has `EIT_UNKNOWN` as default so it set it up only when indices are available

Reply:
ok so you hit this codepath only if indices were supplied by glTF?

Reply:
yeah, it is well considered